### ISSUE TITLE
how to inference  single image?

### ISSUE THREAD

**Issue by gewenpulan** [gudovskiy/cflow-ad#11]
great job
i am new about abnomal detection.
what i want to know is that ,in actually scene ，when i have trained one model with bottle class in mvtec,  i give another bottle image, should the model give the class (nomal or abnomal),and location the abnomal area like GT?
could you give  some demo code for that (load a model and predict single image)?

**Comment by gudovskiy** [comment]
@gewenpulan  I guess, you can just set batch size = 1

**Comment by gewenpulan** [comment]
Thanks for soon reply.
Maybe I was not describe clearly, I tried batch size =1, but it does not work.
I delete all the  file in bottle class in Mvtec dataset(broken_large,broken_small,contamination and good), and then copy 000.png to test folder,like below

bottle
|____test
       |____000.png

When I run the main.py , the dataloader  seems not load the image correctly,should I have to make a GT?
I want to predict a new image which model dose not seen before, and I already have the model, and seg_threshold. what should I do ? 
Thanks again.

**Comment by gudovskiy** [comment]
@gewenpulan 1. [dataloader presumes](https://github.com/gudovskiy/cflow-ad/blob/0d39848c988ad1c6b3c9b7f4bad40d52c1a75ceb/custom_datasets/loader.py#L206) that you have a GT file for every image with anomalies. But for images without anomalies, you don't need corresponding GT. So, in your case you can put your image to bottle/test/good/000.png folder.
2. you need to train a model with anomaly-free images in class/train/good and test it with images in class/test/good. Of course, if you don't have GT, you need to set the detection threshold manually and remove threshold-free metrics like AUROC.

**Comment by gewenpulan** [comment]
@gudovskiy  Thanks very much for your detail explanation.
After reading the code carefully，I have achieved my goal. feed the model one image and seg_threshold output the anomaly image.
While I have another question here, the function _test_meta_epoch_ output the fps, on my machine it took 20ms for one image but when I set viz =True and only add function _export_scores_ to get the anomaly image with mask, it took 80ms.
What I want to ask is if I can get mask only  rely on _test_meta_epoch_ ? or I have to add _export_scores_ ?
The function _test_meta_epoch_  return variable  **test_dist**， if  I can  decide nomal/anomal form this variable? I am not quite understand why calculate time cost only on function_test_meta_epoch_  .

**Comment by gudovskiy** [comment]
@gewenpulan --viz option is only to generate images for the paper. Yes, **test_dist** + simple post-processing in [these lines](https://github.com/gudovskiy/cflow-ad/blob/master/train.py#L311-L329) is enough.

**Comment by gewenpulan** [comment]
@gudovskiy  Sorry for late reply.
Thanks , I understand it.
